### PR TITLE
[WIP] rustc_llvm: use rust-bindgen to generate FFI bindings

### PR DESCRIPTION
**Pull request by tamird**
This is my WIP attempt at #17795; note that the majority of the work involves hoisting various definitions from `rustllvm/*.cpp` to `rustllvm/rustllvm.h`, and then mapping various `LLVMFoo` to `Foo` in Rust to match the current exported names.

This doesn't compile since many definitions haven't yet been hoisted, but `librustc_llvm` itself **does** compile, so this should be a decent proof of concept for y'all to give feedback on before I proceed.

I suspect @eddyb will have opinions.

EDIT: compiles now.

### CODE REVIEW

**Comment by tamird** [comment]
It turns out I can feed the .cpp files directly into bindgen, negating the need for most of the manual work of moving things into the header! Let me know if that's preferable.

**Comment by tamird** [comment]
Ah, there's no `libclang` on the builders. Maybe the generated bindings should just be checked in.

**Comment by tamird** [comment]
Anyway, I was able to bootstrap with this locally 🎉

File: src/rustllvm/RustWrapper.cpp [hunk 1]
```diff
@@ -566,8 +570,8 @@ extern "C" void LLVMRustAddModuleFlag(LLVMModuleRef M, const char *Name,
   unwrap(M)->addModuleFlag(Module::Warning, Name, Value);
 }
 
-extern "C" void LLVMRustMetadataAsValue(LLVMContextRef C, LLVMMetadataRef MD) {
-  wrap(MetadataAsValue::get(*unwrap(C), unwrap(MD)));
+extern "C" LLVMValueRef LLVMRustMetadataAsValue(LLVMContextRef C, LLVMMetadataRef MD) {
+  return wrap(MetadataAsValue::get(*unwrap(C), unwrap(MD)));
```
Comment (tamird) [inline src/rustllvm/RustWrapper.cpp]: here's the first mistake caught by this.

**Comment by cuviper** [comment]
> Ah, there's no libclang on the builders. Maybe the generated bindings should just be checked in.

Requiring `libclang` would also make it harder on third-party builders, like distros.  I think it would make sense to have `bindgen` use as an optional feature.  Then yes, either check in the bindings, or just use them as a sanity check on the current bindings.

**Comment by tamird** [comment]
@cuviper how would I plumb the optional feature through? It doesn't look like I can pass `--features rustc_llvm/rebuild_bindings` from the top level cargo invocation - would I have to plumb "rebuild_llvm_bindings" through the dependency tree?

**Comment by cuviper** [comment]
> would I have to plumb "rebuild_llvm_bindings" through the dependency tree?

Hmm, that's annoying.

It could just be a standalone crate in `src/tools/` which performs bindgen.  It would act basically like an early build script for `rustc_llvm` -- Cargo wouldn't know about it, but rustbuild could manage it and build/run it when enabled.  This may even be better overall, since tools don't have to be rebuilt for each stage of the compiler bootstrap.

**Comment by tamird** [comment]
That's actually somewhat more annoying since we'd need to reimplement the
logic that locates `llvm-config`. Any ideas on that?

**Comment by tamird** [comment]
Actually, that may not be so bad. I'm looking into it.

**Comment by tamird** [comment]
Eh, on second thought, I have no idea how bootstrap is organized. Perhaps someone more familiar with it could help with this part?

**Comment by eddyb** [comment]
cc @alexcrichton

**Comment by alexcrichton** [comment]
Hm I think this requires a dependency on libclang, right? I think for that reason alone (unfortunately) I wouldn't want to take this route. If we want to validate the bindings then I think we should set up [ctest](https://github.com/alexcrichton/ctest) which doesn't require any extra system dependencies.

**Comment by tamird** [comment]
I think the suggestion was to have an in-tree tool that regenerates the
bindings, perhaps in addition to ctest? The advantage of having it in
bootstrap is that bootstrap knows the path to llvm-config.

Looking at ctest, it seems we'll need to run that from bootstrap as well
since it also needs to locate llvm-config to get include paths.

So...who can help me with the bootstrap scaffolding?

On Nov 30, 2017 08:54, "Alex Crichton" <notifications@github.com> wrote:

Hm I think this requires a dependency on libclang, right? I think for that
reason alone (unfortunately) I wouldn't want to take this route. If we want
to validate the bindings then I think we should set up ctest
<https://github.com/alexcrichton/ctest> which doesn't require any extra
system dependencies.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/rust-lang/rust/pull/46353#issuecomment-348194173>, or mute
the thread

**Comment by alexcrichton** [comment]
I would personally prefer to do validation rather than generation of bindings here myself, I've found handwritten bindings to be quite easy to maintain over time and it doesn't require contributors to get libclang set up or anything like that.

For rustbuild integration of verification it would probably look like a dedicated whole new test suite, I don't think it would fit into any  of our existing test suites. You'd basically be twiddling with support in `src/bootstrap/check.rs` I believe.

**Comment by shepmaster** [comment]
@tamird We haven't heard from you in over 7 days — will you be able to address the recent feedback about pivoting to a validation perspective?

**Comment by tamird** [comment]
Yeah, haven't had a chance to do it but I plan to.

**Comment by alexcrichton** [comment]
Ok it's been a few weeks now so I'm going to close this, but we can of course reconsider a resubmission!

**Pull request closed without merging**